### PR TITLE
Hires optimal

### PR DESCRIPTION
**Pull request by rcooke-ast**
The first "complete" Keck/HIRES data reduction. This version also includes support for fully automated wavelength calibration with arclines (currently, only for longslit data reduction). The only missing feature of ARMED is the wavelength calibration, which will follow shortly.

This update should be merged after the corresponding PRs in `arclines` and the `PYPIT development suite` have been merged.

To be honest, this update got a little out of hand... most of the corrections are to the updated settings/calibration files... sorry about that...

### CODE REVIEW

**Comment by rcooke-ast** [comment]
The tests fail, but this is not due to this PR. There appears to be an issue with arcoadd.

**Comment by profxj** [comment]
Do you mean to bring in this file:

MF_shane_kast_blue/MasterSensFunc_C_aa.yaml

?

File: pypit/ararc.py [hunk 1]
```diff
@@ -236,34 +238,36 @@ def setup_param(slf, sc, det, fitsdict):
...
         lamps = ['ArI','NeI','HgI','KrI','XeI']  # Should set according to the lamps that were on
         if disperser == '600/7500':
-            arcparam['n_first']=3 # Too much curvature for 1st order
+            arcparam['n_first']=2 # Too much curvature for 1st order
             arcparam['disp']=0.80 # Ang per pixel (unbinned)
-            arcparam['b1']= 1./arcparam['disp']/slf._msarc[det-1].shape[0]
+            arcparam['disp_toler']=0.05 # Tolerance of dispersion
+            arcparam['b1']= 1./arcparam['disp']/slf._msarc[det-1].shape[0] / binspectral
             arcparam['wvmnx'][1] = 11000.
+            arcparam['Nstrong'] = 30
```
Comment (profxj) [inline pypit/ararc.py]: This could be trouble for some of the lower resolution gratings..

Comment (profxj) [reply pypit/ararc.py]: Ah, I see it is grating dependent.  Could still be trouble.  ;)

File: pypit/ararc.py [hunk 2]
```diff
@@ -340,6 +344,10 @@ def simple_calib(slf, det, get_poly=False):
     # Read Arc linelist
     llist = aparm['llist']
 
+    # np.savetxt("lines.wav", llist['wave'].data)
```
Comment (profxj) [inline pypit/ararc.py]: remove?

File: pypit/arpca.py [hunk 3]
```diff
@@ -123,10 +126,7 @@ def basis(xfit, yfit, coeff, npc, pnpc, weights=None, skipx0=True, x0in=None, ma
     x3fit = np.dot(eigv,high_order_matrix.T) + np.outer(x0fit,np.ones(nrow)).T
     outpar = dict({'high_fit': high_fit, 'x0': x0, 'x0in': x0in, 'x0fit': x0fit, 'x0res': x0res, 'x0mask': fitmask,
                    'hidden': hidden, 'usetrc': usetrace, 'eigv': eigv, 'npc': npc, 'coeffstr': coeffstr})
-    if retmask:
-        return x3fit, outpar, tmask
-    else:
-        return x3fit, outpar
+    return x3fit, outpar
```
Comment (profxj) [inline pypit/arpca.py]: Are you sure that retmask option is no longer needed?

Just checking..

File: pypit/arqa.py [hunk 4]
```diff
@@ -670,24 +670,24 @@ def slit_profile(slf, mstrace, model, lordloc, rordloc, msordloc, textplt="Slit"
                 continue
             spatval = (word[1] + 0.5 - lordloc[:, ndone+j][word[0]]) / (rordloc[:, ndone+j][word[0]] - lordloc[:, ndone+j][word[0]])
             fluxval = mstrace[word]
-            mxval = np.max(fluxval)
+            mxval = 1.25
```
Comment (profxj) [inline pypit/arqa.py]: I'll trust that this is a good idea.

File: pypit/artrace.py [hunk 5]
```diff
@@ -103,13 +104,15 @@ def assign_slits(binarr, edgearr, ednum=100000, lor=-1):
             arrcen = smedgehist[2:-2]
             arrrgt = smedgehist[3:-1]
             arrrfr = smedgehist[4:]
+#            wpk = np.where((arrcen >= arrlft) & (arrcen > arrrgt) &  # Exactly one of these should be >=
```
Comment (profxj) [inline pypit/artrace.py]: remove?

File: pypit/artrace.py [hunk 6]
```diff
@@ -119,11 +122,12 @@ def assign_slits(binarr, edgearr, ednum=100000, lor=-1):
             if np.all(pedges[:, 1]-pedges[:, 0] == 0):
                 # Remaining peaks have no width
                 break
-            if msgs._debug['trace'] and False:
+            if False:#msgs._debug['trace'] and False:
```
Comment (profxj) [inline pypit/artrace.py]: put back the way it was?

File: pypit/data/settings/settings.baseargflag [hunk 7]
```diff
@@ -59,8 +59,8 @@ arc calibrate IDpixels []            # Manually set the pixels to be identified
 arc calibrate IDwaves []             # Manually set the corresponding ID wavelengths
 arc calibrate nfitpix  5             # Number of pixels to fit when deriving the centroid of the arc lines (an odd number is best)
 arc calibrate lamps None           # name of the ions used for the wavelength calibration
-arc calibrate method simple          # What method should be used to fit the individual arc lines (options are: fit, simple); fit is perhaps the most accurate; simple uses a polynomial fit (to the log of a gaussian), is the fastest and is reliable
-arc calibrate detection 6.0         # How significant should the arc line detections be (in units of a standard deviation)
+arc calibrate method arclines          # What method should be used to fit the individual arc lines (options are: fit, simple); fit is perhaps the most accurate; simple uses a polynomial fit (to the log of a gaussian), is the fastest and is reliable
+arc calibrate detection 1.0         # How significant should the arc line detections be (in units of a standard deviation)
```
Comment (profxj) [inline pypit/data/settings/settings.baseargflag]: 1.0 sounds aggressive..

**Comment by rcooke-ast** [comment]
Good catch on this: `MF_shane_kast_blue/MasterSensFunc_C_aa.yaml` - I did not intend for that, and have removed it now... thanks!

File: pypit/ararc.py (hunk 1 above)
Comment (rcooke-ast) [reply pypit/ararc.py]: OK, I'll change this back. It's not used with the new arclines code. I think I did this while I was trying to "fix" the solution for LRISr.

File: pypit/ararc.py (hunk 2 above)
Comment (rcooke-ast) [reply pypit/ararc.py]: done.

File: pypit/arpca.py (hunk 3 above)
Comment (rcooke-ast) [reply pypit/arpca.py]: It doesn't seem to be called by any other function in the code, so I imagine this is now deprecated.

File: pypit/arqa.py (hunk 4 above)
Comment (rcooke-ast) [reply pypit/arqa.py]: Yep - on line 644 the maximum value is 1.25 (the slit profile is normalized to 1, so this is just a bit above 1).

File: pypit/artrace.py (hunk 5 above)
Comment (rcooke-ast) [reply pypit/artrace.py]: Done.

File: pypit/artrace.py (hunk 6 above)
Comment (rcooke-ast) [reply pypit/artrace.py]: Done!

File: pypit/data/settings/settings.baseargflag (hunk 7 above)
Comment (rcooke-ast) [reply pypit/data/settings/settings.baseargflag]: I've now removed the continuum algorithm and this gives a much better result all around. Therefore, I've changed this back to the original value, but at present, this variable doesn't do anything in the code.

**Comment by profxj** [comment]
This PR is ready now.

Merge after arclines and the Development are A-O-K.

**Comment by rcooke-ast** [comment]
This is enough for this PR - I'll make some small changes to PR #284.

**Pull request merged**